File: Fondamenti_programmazione_II/Python/Esami/Gestione_progetti_ricerca/Gestione.py
import sys
import argparse

from Ricerca import Ricerca
from Innovazione import Innovazione
from Ricercatore import Ricercatore
from Partecipazione import Partecipazione

class Gestione:
    def __init__(self, arg):
        self._arg = arg

    def main(self):
        progetti = []
        ricercatori = []
        progetto_dict = {}
        try:
            f = open("/home/thomas/Documenti/GitHub/University/Python/Esami/Gestione_progetti_ricerca/progetti.txt", "r")
            line = f.readline().strip()
            while(line != ''):
                tok = line.split()
                cod = int(tok[0])
                tipo = tok[1]
                titolo = f.readline().strip()
                coordinatore = f.readline().strip()
                organizzazione = f.readline().strip()
                line = f.readline().strip()
                tok = line.split()
                if(tipo == "ricerca"):
                    argomento = tok[0]
                    partner = int(tok[1])
                    importo = float(f.readline().strip())
                    r = Ricerca(cod, titolo, coordinatore, organizzazione, importo, argomento, partner)
                    progetti.append(r)
                    progetto_dict[cod] = r
                else:
                    aziende = int(tok[0])
                    importo = float(tok[1])
                    i = Innovazione(cod, titolo,coordinatore, organizzazione, importo, aziende)
                    progetti.append(i)
                    progetto_dict[cod] = i

                line = f.readline().strip()
            f.close()
        except IOError as i:
            print(i)
        except Exception as e:
            print(e)

        try:
            f = open("/home/thomas/Documenti/GitHub/University/Python/Esami/Gestione_progetti_ricerca/ricercatori.txt", "r")
            line = f.readline().strip()
            while(line != ''):
                nome = line
                cognome = f.readline().strip()
                r = Ricercatore(cognome, nome)
                ricercatori.append(r)
                line = f.readline()
                # a capo o termine
                while(line != "\n" and line != ''):
                    tok = line.split()
                    codice_progetto = int(tok[0])
                    orario = float(tok[1])
                    progetto = progetto_dict[codice_progetto]
                    partecipazione = Partecipazione(progetto, orario)
                    r.addPartecipazione(partecipazione)
                    line = f.readline()
                line = f.readline().strip()

            f.close()
        except IOError as i:
            print(i)

        # stampa elenco progetti
        print("tipo, titolo, codice, coordinatore, organizzazione, argomento, partner, aziende, import totale in migliaia di euro")

        for p in progetti:
            print(p)

        print("\n")

        # stampa elenco ricercatori
        for r in ricercatori:
            print(r)

        print("\n")

        try:
            cognome = self._arg
            for ricercatore in ricercatori:
                if ricercatore._cognome == cognome:
                    partecipazione = ricercatore.getImpegno_Max()
                    print(ricercatore._nome +"\t"+ ricercatore._cognome +"\t"+ str(partecipazione._orario) +"\t"+ partecipazione._progetto._titolo)

        except IOError as i:
            print(i)


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument('cognome', action='store')
    pa = parser.parse_args()
    g = Gestione(pa.cognome)
    g.main()

File: Fondamenti_programmazione_II/Java/EserciziEsame/GestioneProgettiRicercaProf/Gestione.java
package EserciziEsame.GestioneProgettiRicercaProf;

import java.io.*;
import java.util.*;

public class Gestione {

    static List<Progetto> progetti=new LinkedList<Progetto>();
    static List<Ricercatore> ricercatori=new LinkedList<Ricercatore>();
    static Map<Integer,Progetto> codProg=new HashMap<Integer,Progetto>();   

    public static void main(String[] args){

        // Punto 1: memorizzazione dei progetti
        try{

            BufferedReader br=new BufferedReader(new FileReader("progetti.txt"));
            String line=br.readLine();

            while(line!=null){

                StringTokenizer tok=new StringTokenizer(line);
                int codice=Integer.parseInt(tok.nextToken());
                String tipo=tok.nextToken();

                String titolo=br.readLine();

                String coordinatore=br.readLine();

                String organizzazione=br.readLine();

                line=br.readLine();

                if(tipo.equals("ricerca")){

                    tok=new StringTokenizer(line);
                    String argomento=tok.nextToken();
                    int partner=Integer.parseInt(tok.nextToken());

                    double importo=Double.parseDouble(br.readLine());

                    Ricerca r=new Ricerca(codice,titolo,coordinatore,organizzazione,argomento,partner,importo);
                    progetti.add(r);
                    codProg.put(codice,r);
                }
                else{

                    tok=new StringTokenizer(line);
                    int aziende=Integer.parseInt(tok.nextToken());
                    double importo=Double.parseDouble(tok.nextToken());

                    Innovazione i=new Innovazione(codice,titolo,coordinatore,organizzazione,aziende,importo);
                    progetti.add(i);
                    codProg.put(codice,i);
                }

                line=br.readLine();
            }
            br.close();
        }
        catch(IOException e){
            System.err.println(e);
        }
        catch(Exception e){
            System.err.println(e);
        }

        // Punto 2: memorizzazione dei ricercatori
        try{

            BufferedReader br=new BufferedReader(new FileReader("riceratori.txt"));
            String line=br.readLine();

            while(line!=null){

                String nome=line;

                String cognome=br.readLine();

                Ricercatore r=new Ricercatore(nome,cognome);

                ricercatori.add(r);
                line=br.readLine();

                while(line!=null && !line.equals("")){

                    StringTokenizer tok=new StringTokenizer(line);
                    int codice=Integer.parseInt(tok.nextToken());
                    double impegno=Double.parseDouble(tok.nextToken());

                    r.addProgetto(codProg.get(codice), impegno);
                    line=br.readLine();
                }
                line=br.readLine();
            }
            br.close();
        }
        catch(IOException e){
            System.err.println(e);
        }
        catch(Exception e){
            System.err.println(e);
        }

        // Punto 3: visualizzazione di tutti i progetti
        System.out.println("tipo, titolo, codice, coordinatore, organizzazione, argomento, partner, aziende, importo totale in migliaia di euro");

        for(Progetto p:progetti)
            System.out.println(p);

        System.out.println();

        // Punto 4: visualizzazione di tutti i ricercatori
        for(Ricercatore r:ricercatori)
            System.out.println(r);
            
        System.out.println();

        // Punto 5: visualizzazione del progetto con l'impegno orario più alto dato il cognome
        String cognome=args[0];
        for(Ricercatore r:ricercatori){
            if(r.getCognome().equals(cognome)){
                Impegno i=r.getMax();
                System.out.println(r.getNome()+" "+r.getCognome()+" "+i.getOre()+" "+i.getProgetto().getTitolo());
            }
        }
    }
}
